provider/oci/common/client.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package common
+
+import (
+	"context"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/instance"
+	"github.com/juju/juju/network"
+
+	ociCommon "github.com/oracle/oci-go-sdk/common"
+	ociCore "github.com/oracle/oci-go-sdk/core"
+	ociIdentity "github.com/oracle/oci-go-sdk/identity"
+)
+
+type jujuConfigProvider struct {
+	key            []byte
+	keyFingerprint string
+	passphrase     string
+	tenancyOCID    string
+	userOCID       string
+	region         string
+}
+
+type ociClient struct {
+	ociCore.ComputeClient
+	ociCore.BlockstorageClient
+	ociCore.VirtualNetworkClient
+	ociIdentity.IdentityClient
+
+	ociCommon.ConfigurationProvider
+}
+
+// NewJujuConfigProvider returns a new ociCommon.ConfigurationProvider instance
+func NewJujuConfigProvider(user, tenant string, key []byte, fingerprint, passphrase, region string) ociCommon.ConfigurationProvider {
+	return &jujuConfigProvider{
+		key:            key,
+		keyFingerprint: fingerprint,
+		passphrase:     passphrase,
+		tenancyOCID:    tenant,
+		userOCID:       user,
+		region:         region,
+	}
+}
+
+func NewOciClient(provider ociCommon.ConfigurationProvider) (ApiClient, error) {
+	computeClient, err := ociCore.NewComputeClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	blockStorage, err := ociCore.NewBlockstorageClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	virtualNetwork, err := ociCore.NewVirtualNetworkClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	ident, err := ociIdentity.NewIdentityClientWithConfigurationProvider(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &ociClient{
+		computeClient,
+		blockStorage,
+		virtualNetwork,
+		ident,
+		provider,
+	}, nil
+}
+
+// Ping validates that the client can access the OCI API successfully
+func (o *ociClient) Ping() error {
+	tenancyID, err := o.TenancyOCID()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	request := ociIdentity.ListCompartmentsRequest{
+		CompartmentId: &tenancyID,
+	}
+	ctx := context.Background()
```

Comment:
The general recommendation seems to be that context.TODO is better than context.Background for libraries to use.
Thoughts?
At least the go docs seem to say that context.Background is intended for the main() function of an application.

We should probably have a general discussion about how we want to use/expose context.Context objects for the oracle provider, since their library seems to use it extensively.

Reply:
I seem to be wrong with this. I've been looking at other libraries and they do seem to all fall back to context.Background when they don't already have a context to pass through. We should still talk about this internally, about how we want to be handling this.